File: Algorithms/С++/Medium/BubbleSortTwo.cpp

// a kind of bubble sort
// O(n)
void ShakeSort(std::vector<int>& vec) {
	if (vec.empty())
		return;

	int right = vec.size() - 1;
	int left = 0;
	while (left <= right) {
		for (int i = right; i > left; i--) {
			if (vec[i - 1] > vec[i]) {
				swap(vec[i - 1], vec[i]);
			}
		}
		left++;
		for (int i = left; i < right; i++) {
			if (vec[i + 1] < vec[i]) {
				swap(vec[i], vec[i + 1]);
			}
		}
		right--;
	}
}

File: Algorithms/Golang/Medium/BubbleSortTwo.go



// O(n)
func ShakeSort(nums []int) {
	right, left := 0, len(nums)-1

	for left <= right {
		for i := right; i > left; i-- {
			if nums[i] < nums[i-1] {
				temp := nums[i-1]
				nums[i-1] = nums[i]
				nums[i] = temp
			}
		}
		left++

		for j := left; j < right; j++ {
			if nums[j] > nums[j+1] {
				temp := nums[j]
				nums[j] = nums[j+1]
				nums[j+1] = temp
			}
		}
		right--
	}
}
